### PR TITLE
Specify overwrite mechanism for form methods

### PR DESCRIPTION
**Pull request by erdlet**
Hi all,

I just had a little bit of time to begin with the specification of HTML form method overwrites. We already have this mechanism in Krazo, so I used the implementation there as inspiration for the default implementation in the spec. As this is my first addition to the spec, I'm pretty sure there are a lot of things I can do better.

Please let me know what you think about this draft @eclipse-ee4j/ee4j-mvc-committers and anyone else interested.

### CODE REVIEW

**Review by ivargrimstad — APPROVED** [review]
LGTM

File: api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java
Comment (chkal) [inline api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java:38]: Two comments:

1. Does anyone know if there is some kind of "de-facto standard" for the name of this parameter? AFAIK other frameworks support this mechanism as well and if there is some kind of standard name, it would be cool to use it.
2. Maybe we should allow to customize the name of the form field via some configuration parameter?

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (chkal) [inline spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:15]: Is there a `<` missing?

**Review by chkal — COMMENTED** [review]
@erdlet Thank you very much for working on this. I would really love to see this in the spec.

However, I want to share some thoughts. To be honest, I'm not completely sure if we really need this SPI interface which people can implement to provide custom behavior. Isn't the default implementation everything people actually require in real work projects?

When I thought about this feature before, I was actually thinking about just providing:

* A configuration parameter which enables or disables the form method override feature-
* A configuration parameter to customize the name of the request parameter which contains the HTTP verb. 

And that's all. Isn't this sufficient for 99% of all use cases?

And if people really want to have some kind of fancy custom behavior, they can actually implement it with a standard JAX-RS `ContainerRequestFilter`, which has just a few lines of code. Even the default mechanism which MVC implementations provide boils down to something like this:

```
@Provider
@PreMatching
public class MethodContainerRequestFilter implements ContainerRequestFilter {

  @Override
  public void filter( ContainerRequestContext requestContext ) throws IOException {

    String method = requestContext.getHeaderString( "X-Method" );
    if( method != null ) {
      requestContext.setMethod( method );
    }

  }

}
```

So I wonder if this SPI isn't an overkill for this *simple feature*?

@eclipse-ee4j/ee4j-mvc-committers Any other thoughts?

**Comment by erdlet** [comment]
Thanks for your feedback @chkal. After thinking a few days about that proposal I came to the same conclusion that the SPI would be an overkill. 

So I think I'll change the proposal so that only the current default mechanism is required as well as the configuration parameters.

**Comment by chkal** [comment]
> So I think I'll change the proposal so that only the current default mechanism is required as well as the configuration parameters.

+1 from my side.

@eclipse-ee4j/ee4j-mvc-committers It would be great to hear other thoughts as well.

File: api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java
Comment (erdlet) [reply api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java:38]: All frameworks I know use `_method` (especially Spring), so this name should be the `de-facto` standard. I added a possibility to overwrite this name in my new draft.

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (erdlet) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:15]: Yes. Thanks for that catch :)

**Comment by erdlet** [comment]
Hi @eclipse-ee4j/ee4j-mvc-committers,

I changed the proposal according to my and @chkal's thoughts. I think this draft is enough to start and will cover 99% of the common use-cases. Would be great to get some feedback.

@jelemux Maybe you want to give some feedback too?

**Comment by jelemux** [comment]
So if I understand correctly, we can have a form which sends a `POST` but because of the hidden input MVC interprets it for example like a `PATCH`. Therefore a form like that could call a controller with a `@PATCH` method. Is this correct?
If so, this looks good to me.

Will there be TCK tests this?

**Comment by erdlet** [comment]
Thanks @jelemux for your feedback!

> So if I understand correctly, we can have a form which sends a POST but because of the hidden input MVC interprets it for example like a PATCH. Therefore a form like that could call a controller with a @PATCH method. Is this correct?

Absolutely correct 👍

> Will there be TCK tests for this?

Yes, I just want to have a final draft for this before I start to work on the tests to avoid double work in case of changes. I plan to test the behavior of the config options and usage of a different hidden field name. Think it will sum up to three or four tests.

**Review by jelemux — APPROVED** [review]

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (erdlet) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:15]: Good point, I'll change it.

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (erdlet) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:40]: Yes, the user MUST use a `POST`, as all alternative HTTP methods provide write access and accessing their semantic is our target. Therefore `GET` is out of scope as this is a read-only function.

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (erdlet) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:41]: TBH I thought in servlet terms as I know that the `request attribute` is the place where form parameters are mapped to. But you're right, it's better to use your suggestion. Thanks :)

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (erdlet) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:44]: > I'm wondering if the validation of the HTTP verb should really be optional.

It isn't really optional as the underlaying Jakarta REST or Servlet implementation is doing checks on its own but throws an exception in case of some invalid value. But I don't know how people want to handle this case, as some prefer exceptions and some NOOPs. The optional requirement from our side would allow both methods.

On the other hand, maybe it'd be better to be more precise and define exactly what to do. I need to think a little bit about that. Maybe someone else wants to share their opinion too.

File: api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java
Comment (erdlet) [reply api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java:26]: > Is it really just about HTML forms?

My first impulse is to say "Yes", as JavaScript clients could target the API by using the correct method directly. But that depends, as always, on the client developer. So personally I'm not attached to the _form_ inside the class name.

> Maybe we could use overwrite instead of overwriter?

Yes, absolutely. This was a leftover of the initial draft.

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (chkal) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:40]: Ok, makes sense. `GET` is idempotent and this shouldn't be changed by the new mechanism.

File: api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java
Comment (chkal) [reply api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java:26]: > My first impulse is to say "Yes", as JavaScript clients could target the API by using the correct method directly. But that depends, as always, on the client developer. So personally I'm not attached to the form inside the class name.

After thinking a bit more about this: Actually, even in case of PPR requests, the request would submit a form, although from JavaScript using `FormData` or some other mechanism. On the other hand: In theory, an MVC Controller method could also process some kind of JSON document sent via POST and return a view in HTML, or am I missing something?

**Comment by erdlet** [comment]
Hi all,

I'm afraid there won't be more feedback. So if this PR is OK for the reviewers, it would be great when @chkal or @ivargrimstad merge it. After that I'll implement the TCK tests and adapt Krazo core.

File: spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc
Comment (chkal) [reply spec/src/main/asciidoc/chapters/form-method-overwrite.asciidoc:44]: I'm also still not sure whether we should do the validation or not. Maybe we should simply don't mention validation at all for now. This will still allow us to define the exact behavior later on!? So my proposal is to just remove this sentence:

```suggestion
.. Overwrite the HTTP method to the value provided by the hidden form field.
```

**Review by chkal — COMMENTED** [review]
I added a few more comments. I also included proposed changes for all my comments to make it easier to apply them if we go with them.

Sorry for the delay in my review and for even more comments. But I guess it would be preferable to have it polished before merging it. :-)

File: api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java
Comment (erdlet) [reply api/src/main/java/jakarta/mvc/form/FormMethodOverwriter.java:20]: yes, you're absolutely right. Missed that sentence.

**Comment by erdlet** [comment]
@chkal thanks for the feedback. I hope the spec chapter is fine now :)

**Comment by chkal** [comment]
Awesome! Thanks a lot for applying the changes.

One last thing. I think the new chapter is still missing in the `mvc-spec.asciidoc` file and therefore not included in the resulting PDF, or am I missing something. Sorry, just noticed this while looking at all the changes in the PR.

**Comment by erdlet** [comment]
Thanks @chkal, I really missed to add the new chapter :+1:

**Review by chkal — APPROVED** [review]
Awesome! Thanks!

@ivargrimstad @jelemux Could you check the PR again and approve? Thanks!

**Comment by chkal** [comment]
@ivargrimstad @jelemux Ah, sorry. Looks like your approval is still valid. Never mind. :-)

**Pull request merged**
